### PR TITLE
Deterministic Builds

### PR DESCRIPTION
**Pull request by jaredpar**
This PR changes the Roslyn repro in the following ways:

- Builds now take advantage of the `/deterministic` build switch ideally producing identical byte for byte DLLs between equal builds.  
- Test for verifying Roslyn builds deterministicly against itself: `cibuild.cmd /testDeterminism`

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-compiler, @dotnet/roslyn-infrastructure

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 1]
```diff
@@ -58,6 +58,13 @@ public CommonCompiler(CommandLineParser parser, string responseFile, string[] ar
             this.Arguments = parser.Parse(allArgs, baseDirectory, sdkDirectoryOpt, additionalReferenceDirectories);
             this.MessageProvider = parser.MessageProvider;
             this.AnalyzerLoader = analyzerLoader;
+
+#if DEBUG
+            if (Arguments.ParseOptions.Features.ContainsKey("debug-determinism"))
+            {
+                EmitDeterminismKey(Arguments, args, baseDirectory, parser);
+            }
+#endif
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: This is a block that I'm interested in getting feedback on.  

From personal experience I can attest that it's nearly impossible to diagnose deterministic build issues without a diagnostic file of this nature.  There are just too many subtle factors that can affect determinism and it's impossible to verify them all in the debugger.  Having a file get emitted to disk which captures a summary of the build makes it super easy to compare builds: just diff the diagnostic file.  

Eventually I'd like to have this be a feature for customers to consume as well.  I'm not sure this is the best approach though.

Comment (TyOverby) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: +1 on this from me.  Should this be a feature or a fully supported flag?

**Comment by jaredpar** [comment]
Updated to work around the last known determinism bug.  This allowed me to extend testing to all of Roslyn.sln.  

CC @dotnet/roslyn-infrastructure, @dotnet/roslyn-compiler

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: I think it is fine to have this as an "unsupported" feature flag on all (not just Debug) builds. If it becomes popular or is demonstrated useful to external customers, we can consider exposing it as a supported option, perhaps modified or enhanced per feedback.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 2]
```diff
@@ -58,6 +58,13 @@ public CommonCompiler(CommandLineParser parser, string responseFile, string[] ar
             this.Arguments = parser.Parse(allArgs, baseDirectory, sdkDirectoryOpt, additionalReferenceDirectories);
             this.MessageProvider = parser.MessageProvider;
             this.AnalyzerLoader = analyzerLoader;
+
+#if DEBUG
```
Comment (gafter) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: I don't think there is any reason to condition this on DEBUG.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 3]
```diff
@@ -847,5 +854,65 @@ protected virtual CultureInfo Culture
...
+                    hashValue = data.Replace("-", "");
+                }
+                catch (Exception ex)
+                {
+                    hashValue = $"Could not compute {ex.Message}";
+                }
+                builder.AppendLine($"\t{sourceFileName} - {hashValue}");
+            }
+
+            return builder.ToString();
```
Comment (gafter) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: What about referenced assemblies? If we don't record them, customers will complain we're not deterministic even though the difference is due to different referenced assemblies (e.g. on different build machines).

**Comment by gafter** [comment]
LGTM, though please consider whether or not you want to do anything about my comments.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 2 above)
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: Glad you agree.  This is a bit of a first (using features to implement a diagnostic switch) so I wanted to get feedback here before I settled on this approach.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 3 above)
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: I'll add a comment clarifying this isn't a complete representation yet.  It contained just the information I needed to track down the issues I was seeing in the code base.  The plan though is to evolve this to the complete representation.  

Issue #8193 will track fleshing this out to the full set of information.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 4]
```diff
@@ -58,6 +58,11 @@ public CommonCompiler(CommandLineParser parser, string responseFile, string[] ar
             this.Arguments = parser.Parse(allArgs, baseDirectory, sdkDirectoryOpt, additionalReferenceDirectories);
             this.MessageProvider = parser.MessageProvider;
             this.AnalyzerLoader = analyzerLoader;
+
+            if (Arguments.ParseOptions.Features.ContainsKey("debug-determinism"))
```
Comment (gafter) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: FYI, this is precisely the sort of thing the features flag was designed for.

**Pull request merged**